File: tasks/task-01-collector/app/storage.py
from __future__ import annotations

import json
import re
import shutil
import uuid
from dataclasses import dataclass
from datetime import UTC, datetime
from pathlib import Path
from tempfile import NamedTemporaryFile


DEFAULT_RUN_ID = "default"
_RUN_ID_PATTERN = re.compile(r"^[A-Za-z0-9][A-Za-z0-9._-]{0,127}$")
_INDEX_FILE = "messages-index.json"


@dataclass(frozen=True)
class MessageEntry:
    sequence: int
    payload_path: Path
    header_path: Path
    created_at: str


class FileStorage:
    def __init__(self, root: Path) -> None:
        self.root = root.resolve()
        self.runs_root = self.root / "runs"
        self.current_run_file = self.root / "current_run.txt"

    def ensure_available(self) -> bool:
        try:
            self.runs_root.mkdir(parents=True, exist_ok=True)
            with NamedTemporaryFile(dir=self.root, delete=True):
                pass
            return True
        except OSError:
            return False

    def create_run(self) -> str:
        self.clear_all_runs()
        run_id = uuid.uuid4().hex
        self._create_run_dir(run_id)
        self.set_current_run(run_id)
        return run_id

    def get_current_run(self) -> str | None:
        if not self.current_run_file.exists():
            return None
        run_id = self.current_run_file.read_text(encoding="utf-8").strip()
        if not run_id:
            return None
        try:
            return _validated_run_id(run_id)
        except ValueError:
            return None

    def set_current_run(self, run_id: str | None) -> None:
        if run_id is None:
            if self.current_run_file.exists():
                self.current_run_file.unlink()
            return
        run_id = _validated_run_id(run_id)
        self.root.mkdir(parents=True, exist_ok=True)
        self.current_run_file.write_text(run_id, encoding="utf-8")

    def end_run(self, run_id: str) -> bool:
        run_id = _validated_run_id(run_id)
        run_dir = self._find_run_dir(run_id)
        if run_dir is None:
            return False

        run_meta = run_dir / "run.json"
        if run_meta.exists():
            data = self._read_json(run_meta)
            data["endedAt"] = _now_iso()
            self._write_json(run_meta, data)

        if self.get_current_run() == run_id:
            self.set_current_run(None)
        return True

    def delete_run(self, run_id: str) -> bool:
        run_id = _validated_run_id(run_id)
        run_dir = self._find_run_dir(run_id)
        if run_dir is None:
            return False

        shutil.rmtree(run_dir)
        if self.get_current_run() == run_id:
            self.set_current_run(None)
        return True

    def clear_all_runs(self) -> None:
        if self.runs_root.exists():
            shutil.rmtree(self.runs_root)
        self.runs_root.mkdir(parents=True, exist_ok=True)

    def collect(self, run_id: str, message_id: str, payload: bytes, headers: dict[str, str]) -> dict:
        run_id = _validated_run_id(run_id)
        run_dir = self._ensure_run_dir(run_id)

        message_key, message_dir = self._resolve_message_group(run_dir, message_id)
        sequence = self._reserve_next_sequence(message_dir)

        entry_dir = message_dir / str(sequence)
        entry_dir.mkdir(parents=True, exist_ok=False)

        payload_path = entry_dir / "payload.bin"
        header_path = entry_dir / "headers.json"
        metadata_path = entry_dir / "metadata.json"

        payload_path.write_bytes(payload)
        self._write_json(header_path, headers)

        created_at = _now_iso()
        self._write_json(
            metadata_path,
            {
                "messageId": message_id,
                "messageKey": message_key,
                "sequence": sequence,
                "runId": run_id,
                "createdAt": created_at,
            },
        )

        return {
            "runId": run_id,
            "messageId": message_id,
            "messageKey": message_key,
            "sequence": sequence,
            "storedAt": str(entry_dir.relative_to(self.root)),
            "createdAt": created_at,
        }

    def get_manifest(self, run_id: str) -> dict:
        run_id = _validated_run_id(run_id)
        run_dir = self._find_run_dir(run_id)
        if run_dir is None:
            return {"runId": run_id, "messages": []}

        messages_root = run_dir / "messages"
        if not messages_root.exists():
            return {"runId": run_id, "messages": []}

        messages: list[dict] = []
        for message_dir in sorted(path for path in messages_root.iterdir() if path.is_dir()):
            entries = self._message_entries(message_dir)
            if not entries:
                continue
            message_id = self._read_message_id(entries[0])
            messages.append(
                {
                    "messageId": message_id,
                    "messageKey": message_dir.name,
                    "count": len(entries),
                    "sequences": [entry.sequence for entry in entries],
                    "latestSequence": entries[-1].sequence,
                }
            )

        messages.sort(key=lambda item: item["messageId"])
        return {"runId": run_id, "messages": messages}

    def get_message_group(self, run_id: str, message_id: str) -> dict | None:
        run_id = _validated_run_id(run_id)
        message_dir = self._message_dir_for_id(run_id, message_id)
        if message_dir is None or not message_dir.exists():
            return None

        entries = self._message_entries(message_dir)
        if not entries:
            return None

        return {
            "runId": run_id,
            "messageId": message_id,
            "messageKey": message_dir.name,
            "entries": [
                {
                    "sequence": entry.sequence,
                    "createdAt": entry.created_at,
                    "payloadPath": str(entry.payload_path.relative_to(self.root)),
                    "headerPath": str(entry.header_path.relative_to(self.root)),
                }
                for entry in entries
            ],
        }

    def get_headers(self, run_id: str, message_id: str, sequence: int | None = None) -> dict | None:
        run_id = _validated_run_id(run_id)
        entry = self._get_entry(run_id, message_id, sequence)
        if entry is None:
            return None
        return self._read_json(entry.header_path)

    def get_payload(self, run_id: str, message_id: str, sequence: int | None = None) -> bytes | None:
        run_id = _validated_run_id(run_id)
        entry = self._get_entry(run_id, message_id, sequence)
        if entry is None:
            return None
        return entry.payload_path.read_bytes()

    def release_message(self, run_id: str, message_id: str) -> bool:
        run_id = _validated_run_id(run_id)
        run_dir = self._find_run_dir(run_id)
        if run_dir is None:
            return False

        message_dir = self._message_dir_for_id(run_id, message_id)
        if message_dir is None or not message_dir.exists():
            return False

        shutil.rmtree(message_dir)
        self._remove_message_from_index(run_dir, message_id)
        return True

    def residual(self, run_id: str) -> dict:
        run_id = _validated_run_id(run_id)
        manifest = self.get_manifest(run_id)
        return {
            "runId": run_id,
            "messageGroups": len(manifest["messages"]),
            "entries": sum(item["count"] for item in manifest["messages"]),
            "messages": manifest["messages"],
        }

    def _ensure_run_dir(self, run_id: str) -> Path:
        run_dir = self._find_run_dir(run_id)
        if run_dir is not None:
            return run_dir
        return self._create_run_dir(run_id)

    def _create_run_dir(self, run_id: str) -> Path:
        run_internal_key = uuid.uuid4().hex
        run_dir = self.runs_root / run_internal_key
        run_dir.mkdir(parents=True, exist_ok=False)
        self._write_json(
            run_dir / "run.json",
            {"runId": run_id, "createdAt": _now_iso(), "endedAt": None},
        )
        self._write_json(run_dir / _INDEX_FILE, {})
        return run_dir

    def _find_run_dir(self, run_id: str) -> Path | None:
        if not self.runs_root.exists():
            return None

        for run_dir in self.runs_root.iterdir():
            if not run_dir.is_dir():
                continue
            run_meta = run_dir / "run.json"
            if not run_meta.exists():
                continue
            meta = self._read_json(run_meta)
            if meta.get("runId") == run_id:
                return run_dir
        return None

    def _message_dir_for_id(self, run_id: str, message_id: str) -> Path | None:
        run_dir = self._find_run_dir(run_id)
        if run_dir is None:
            return None
        mapping = self._read_index(run_dir)
        message_key = mapping.get(message_id)
        if not message_key:
            return None
        return run_dir / "messages" / message_key

    def _resolve_message_group(self, run_dir: Path, message_id: str) -> tuple[str, Path]:
        mapping = self._read_index(run_dir)
        message_key = mapping.get(message_id)
        if message_key is None:
            message_key = uuid.uuid4().hex
            mapping[message_id] = message_key
            self._write_index(run_dir, mapping)

        message_dir = run_dir / "messages" / message_key
        message_dir.mkdir(parents=True, exist_ok=True)
        return message_key, message_dir

    def _reserve_next_sequence(self, message_dir: Path) -> int:
        existing_sequences = [int(path.name) for path in message_dir.iterdir() if path.is_dir() and path.name.isdigit()]
        next_sequence = max(existing_sequences, default=0) + 1
        while (message_dir / str(next_sequence)).exists():
            next_sequence += 1
        return next_sequence

    def _get_entry(self, run_id: str, message_id: str, sequence: int | None = None) -> MessageEntry | None:
        message_dir = self._message_dir_for_id(run_id, message_id)
        if message_dir is None or not message_dir.exists():
            return None

        entries = self._message_entries(message_dir)
        if not entries:
            return None

        if sequence is None:
            return entries[-1]
        for entry in entries:
            if entry.sequence == sequence:
                return entry
        return None

    def _message_entries(self, message_dir: Path) -> list[MessageEntry]:
        entries: list[MessageEntry] = []

        for entry_dir in message_dir.iterdir():
            if not entry_dir.is_dir() or not entry_dir.name.isdigit():
                continue

            metadata_file = entry_dir / "metadata.json"
            if not metadata_file.exists():
                continue

            meta = self._read_json(metadata_file)
            payload_path = entry_dir / "payload.bin"
            header_path = entry_dir / "headers.json"

            if not payload_path.exists() or not header_path.exists():
                continue

            entries.append(
                MessageEntry(
                    sequence=int(entry_dir.name),
                    payload_path=payload_path,
                    header_path=header_path,
                    created_at=str(meta.get("createdAt", "")),
                )
            )

        entries.sort(key=lambda item: item.sequence)
        return entries

    def _read_message_id(self, entry: MessageEntry) -> str:
        metadata_path = entry.payload_path.parent / "metadata.json"
        meta = self._read_json(metadata_path)
        return str(meta.get("messageId", ""))

    def _read_index(self, run_dir: Path) -> dict[str, str]:
        index_file = run_dir / _INDEX_FILE
        if not index_file.exists():
            return {}
        content = self._read_json(index_file)
        return {str(key): str(value) for key, value in content.items()}

    def _write_index(self, run_dir: Path, mapping: dict[str, str]) -> None:
        self._write_json(run_dir / _INDEX_FILE, mapping)

    def _remove_message_from_index(self, run_dir: Path, message_id: str) -> None:
        mapping = self._read_index(run_dir)
        if message_id in mapping:
            mapping.pop(message_id)
            self._write_index(run_dir, mapping)

    @staticmethod
    def _write_json(path: Path, data: dict) -> None:
        path.write_text(json.dumps(data, ensure_ascii=False, separators=(",", ":")), encoding="utf-8")

    @staticmethod
    def _read_json(path: Path) -> dict:
        return json.loads(path.read_text(encoding="utf-8"))


def _validated_run_id(run_id: str) -> str:
    if not _RUN_ID_PATTERN.fullmatch(run_id):
        raise ValueError("Invalid run id")
    return run_id


def _now_iso() -> str:
    return datetime.now(UTC).isoformat()

File: tasks/task-01-collector/src/storage.js
const fs = require('node:fs/promises');
const path = require('node:path');
const { randomUUID } = require('node:crypto');

const DEFAULT_RUN = 'default';

class FileCollectorStorage {
  constructor(options = {}) {
    this.dataDir = options.dataDir || path.resolve(process.cwd(), 'data');
    this.currentRunFile = path.join(this.dataDir, 'current-run.json');
    this.locks = new Map();
  }

  async init() {
    await fs.mkdir(this.dataDir, { recursive: true });
  }

  async health() {
    await this.init();
    const probeDir = path.join(this.dataDir, '.health');
    await fs.mkdir(probeDir, { recursive: true });
    await fs.rm(probeDir, { recursive: true, force: true });
    return { status: 'UP', dataDir: this.dataDir };
  }

  async createRun(runId = randomUUID()) {
    await this.init();
    const manifest = this.#emptyManifest(runId);
    await this.#writeManifest(runId, manifest);
    await this.#writeCurrentRun(runId);
    return { runId };
  }

  async getCurrentRun() {
    try {
      const content = await fs.readFile(this.currentRunFile, 'utf8');
      return JSON.parse(content);
    } catch (error) {
      if (error.code === 'ENOENT') {
        return null;
      }
      throw error;
    }
  }

  async endRun(runId) {
    const currentRun = await this.getCurrentRun();
    if (currentRun && currentRun.runId === runId) {
      await fs.rm(this.currentRunFile, { force: true });
    }
    return { runId, ended: true };
  }

  async deleteRun(runId) {
    await fs.rm(this.#runDir(runId), { recursive: true, force: true });
    const currentRun = await this.getCurrentRun();
    if (currentRun && currentRun.runId === runId) {
      await fs.rm(this.currentRunFile, { force: true });
    }
  }

  async collect({ runId, messageId, payload, headers, contentType }) {
    if (!messageId) {
      const error = new Error('Missing messageId');
      error.statusCode = 400;
      throw error;
    }

    const effectiveRunId = runId || (await this.getCurrentRun())?.runId || DEFAULT_RUN;
    const lockKey = `${effectiveRunId}:${messageId}`;
    return this.#withLock(lockKey, async () => {
      await this.init();
      const manifest = await this.#readManifest(effectiveRunId);
      const existing = manifest.messages[messageId] || [];
      const sequenceNumber = existing.length + 1;
      const messageDir = this.#messageDir(effectiveRunId, messageId);
      await fs.mkdir(messageDir, { recursive: true });
      const payloadFile = path.join(messageDir, `${sequenceNumber}.payload`);
      const headerFile = path.join(messageDir, `${sequenceNumber}.headers.json`);
      await fs.writeFile(payloadFile, payload);
      await fs.writeFile(headerFile, JSON.stringify(headers, null, 2));

      const entry = {
        sequenceNumber,
        contentType,
        storedAt: new Date().toISOString(),
        payloadPath: path.relative(this.#runDir(effectiveRunId), payloadFile).replace(/\\/g, '/'),
        headerPath: path.relative(this.#runDir(effectiveRunId), headerFile).replace(/\\/g, '/'),
      };
      manifest.messages[messageId] = existing.concat(entry);
      manifest.updatedAt = new Date().toISOString();
      await this.#writeManifest(effectiveRunId, manifest);
      return {
        runId: effectiveRunId,
        messageId,
        sequenceNumber,
        contentType,
        location: `/runs/${encodeURIComponent(effectiveRunId)}/messages/${encodeURIComponent(messageId)}`,
      };
    });
  }

  async getRunMessages(runId) {
    const manifest = await this.#readManifest(runId);
    return {
      runId,
      messages: Object.entries(manifest.messages)
        .sort(([left], [right]) => left.localeCompare(right))
        .map(([messageId, documents]) => ({ messageId, documents })),
      releasedMessageIds: [...manifest.releasedMessageIds].sort(),
      createdAt: manifest.createdAt,
      updatedAt: manifest.updatedAt,
    };
  }

  async getMessage(runId, messageId) {
    const manifest = await this.#readManifest(runId);
    const documents = manifest.messages[messageId];
    if (!documents) {
      return null;
    }
    return { runId, messageId, documents };
  }

  async getPayload(runId, messageId, sequenceNumber) {
    const document = await this.#getDocument(runId, messageId, sequenceNumber);
    return {
      contentType: document.contentType,
      payload: await fs.readFile(path.join(this.#runDir(runId), document.payloadPath)),
    };
  }

  async getHeader(runId, messageId, sequenceNumber) {
    const document = await this.#getDocument(runId, messageId, sequenceNumber);
    const content = await fs.readFile(path.join(this.#runDir(runId), document.headerPath), 'utf8');
    return JSON.parse(content);
  }

  async releaseMessage(runId, messageId) {
    const manifest = await this.#readManifest(runId);
    if (!manifest.messages[messageId]) {
      return false;
    }
    delete manifest.messages[messageId];
    manifest.releasedMessageIds = Array.from(new Set([...manifest.releasedMessageIds, messageId]));
    manifest.updatedAt = new Date().toISOString();
    await fs.rm(this.#messageDir(runId, messageId), { recursive: true, force: true });
    await this.#writeManifest(runId, manifest);
    return true;
  }

  async getResidual(runId) {
    const manifest = await this.#readManifest(runId);
    return {
      runId,
      residualMessages: Object.entries(manifest.messages)
        .sort(([left], [right]) => left.localeCompare(right))
        .map(([messageId, documents]) => ({
          messageId,
          documentCount: documents.length,
          sequenceNumbers: documents.map((document) => document.sequenceNumber),
        })),
      releasedMessageIds: [...manifest.releasedMessageIds].sort(),
    };
  }

  async #getDocument(runId, messageId, sequenceNumber) {
    const message = await this.getMessage(runId, messageId);
    if (!message) {
      return null;
    }
    const document = message.documents.find((entry) => entry.sequenceNumber === sequenceNumber) || message.documents[0];
    if (!document) {
      return null;
    }
    return document;
  }

  #runDir(runId) {
    return path.join(this.dataDir, runId);
  }

  #messageDir(runId, messageId) {
    return path.join(this.#runDir(runId), encodeURIComponent(messageId));
  }

  async #readManifest(runId) {
    const runDir = this.#runDir(runId);
    await fs.mkdir(runDir, { recursive: true });
    const manifestFile = path.join(runDir, 'manifest.json');
    try {
      const content = await fs.readFile(manifestFile, 'utf8');
      const parsed = JSON.parse(content);
      parsed.messages ||= {};
      parsed.releasedMessageIds ||= [];
      return parsed;
    } catch (error) {
      if (error.code !== 'ENOENT') {
        throw error;
      }
      const manifest = this.#emptyManifest(runId);
      await this.#writeManifest(runId, manifest);
      return manifest;
    }
  }

  async #writeManifest(runId, manifest) {
    const runDir = this.#runDir(runId);
    await fs.mkdir(runDir, { recursive: true });
    await fs.writeFile(path.join(runDir, 'manifest.json'), JSON.stringify(manifest, null, 2));
  }

  async #writeCurrentRun(runId) {
    await fs.writeFile(this.currentRunFile, JSON.stringify({ runId }, null, 2));
  }

  #emptyManifest(runId) {
    const now = new Date().toISOString();
    return {
      runId,
      createdAt: now,
      updatedAt: now,
      releasedMessageIds: [],
      messages: {},
    };
  }

  async #withLock(key, task) {
    const previous = this.locks.get(key) || Promise.resolve();
    const next = previous.then(task, task);
    this.locks.set(key, next.catch(() => {}));
    try {
      return await next;
    } finally {
      if (this.locks.get(key) === next) {
        this.locks.delete(key);
      }
    }
  }
}

module.exports = {
  FileCollectorStorage,
};
